include/ginkgo/core/multigrid/amgx_pgm.hpp

```diff
@@ -0,0 +1,162 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_MULTIGRID_AMGX_PGM_HPP_
+#define GKO_CORE_MULTIGRID_AMGX_PGM_HPP_
+
+
+#include <vector>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/multigrid/restrict_prolong.hpp>
+
+
+namespace gko {
+namespace multigrid {
+
+
+/**
+ * Amgx parallel graph match (AmgxPgm) is the aggregate method introduced in the
+ * paper M. Naumov et al., "AmgX: A Library for GPU Accelerated Algebraic
+ * Multigrid and Preconditioned Iterative Methods". Current implemenation only
+ * contains size = 2 version.
+ *
+ * AmgxPgm creates the aggreagate group according to the matrix value not the
+ * structure.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes
+ *
+ * @ingroup RestrictProlong
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class AmgxPgm : public EnableRestrictProlong<AmgxPgm<ValueType, IndexType>> {
+    friend class EnableRestrictProlong<AmgxPgm>;
+    friend class EnablePolymorphicObject<AmgxPgm, RestrictProlong>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Returns the system operator (matrix) of the linear system.
+     *
+     * @return the system operator (matrix)
+     */
+    std::shared_ptr<const LinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
+
+    /**
+     * Returns the aggregate group.
+     *
+     * Aggregate group whose size is same as the number of rows stores the
+     * mapping information from row index to coarse row index.
+     * i.e., agg[row_idx] = coarse_row_idx.
+     *
+     * @return the aggregate group.
+     */
+    IndexType *get_agg() noexcept { return agg_.get_data(); }
+
+    /**
+     * @copydoc AmgxPgm::get_agg()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const IndexType *get_const_agg() const noexcept
+    {
+        return agg_.get_const_data();
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * The maximum number of iteration.
+         */
+        unsigned GKO_FACTORY_PARAMETER(max_iterations, 15u);
+
+        /**
+         * The maximum ratio of unassigned number.
+         */
+        double GKO_FACTORY_PARAMETER(max_unassigned_percentage, 0.05);
+
+        /**
+         * Use the deterministic assign_to_exist_agg method or not.
+         */
+        bool GKO_FACTORY_PARAMETER(deterministic, false);
```

Comment:
maybe you can also explain in short, the effects of using deterministic or non-deterministic ?